hudi-aws/src/main/java/org/apache/hudi/aws/transaction/lock/S3StorageLockClient.java

```diff
@@ -0,0 +1,248 @@
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.regions.providers.DefaultAwsRegionProviderChain;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Properties;
+import java.util.function.Supplier;
+
+/**
+ * S3-based distributed lock service using ETag checks (AWS SDK v2).
+ * See RFC: <a href="https://github.com/apache/hudi/blob/master/rfc/rfc-91/rfc-91.md">RFC-91</a>
+ */
+@ThreadSafe
+public class S3StorageLockClient implements StorageLock {
+
+  private static final Logger LOG = LoggerFactory.getLogger(S3StorageLockClient.class);
+  private static final long WAIT_TIME_FOR_RETRY_MS = 1000L;
+  private static final int PRECONDITION_FAILURE_ERROR_CODE = 412;
+  private static final int NOT_FOUND_ERROR_CODE = 404;
+  private static final int CONDITIONAL_REQUEST_CONFLICT_ERROR_CODE = 409;
+  private static final int RATE_LIMIT_ERROR_CODE = 429;
+  private static final int INTERNAL_SERVER_ERROR_CODE_MIN = 500;
+
+  private final Logger logger;
+  private final S3Client s3Client;
+  private final String bucketName;
+  private final String lockFilePath;
+  private final String ownerId;
+
+  /** Constructor that is used by reflection to instantiate an S3-based storage locking client.
+   * @param ownerId The owner id.
+   * @param bucketName The name of the bucket.
+   * @param lockFilePath The path within the bucket where to write lock files.
+   * @param props The properties for the lock config, can be used to customize client.
+   */
+  public S3StorageLockClient(String ownerId, String bucketName, String lockFilePath, Properties props) {
+    this(
+            ownerId,
+            bucketName,
+            lockFilePath,
+            createDefaultS3Client(props),
+            LOG);
+  }
+
+  @VisibleForTesting
+  S3StorageLockClient(String ownerId,
+                                String bucketName,
+                                String lockFilePath,
+                                S3Client s3Client,
+                                Logger logger) {
+    this.s3Client = s3Client;
+    this.lockFilePath = lockFilePath;
+    this.bucketName = bucketName;
+    this.ownerId = ownerId;
+    this.logger = logger;
+  }
+
+  private static S3Client createDefaultS3Client(Properties props) {
+    Region region;
+    try {
+      region = DefaultAwsRegionProviderChain.builder().build().getRegion();
+    } catch (SdkClientException e) {
+      // Fallback to us-east-1 if no region is found
+      region = Region.US_EAST_1;
```

Comment:
lets pull into a constant.. or should we error out instead if no region is found. 

i.e is there a valid case, where no region is found, but US_EAST_1 works? Like to more this fallback, if we think it should 99% not fail

Reply:
Yeah this is a new AWS functionality with regions. Before you could have a default region, but now you have to explicitly specify the region. If the region is not specified somewhere in the environment, then we should default to something like US EAST 1.

https://repost.aws/questions/QUMZwhOFYDQvGsFgdoa6nWlg/regions-getcurrentregion-for-aws-sdk-2-0

Reply:
Yeah I think this is the proper way to do it. If someone complains that the lock provider loading is taking too long then they need to set the region environment variable for their hudi table.